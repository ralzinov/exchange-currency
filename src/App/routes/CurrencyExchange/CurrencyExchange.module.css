.host {
    display: flex;
    flex: 1;
    flex-direction: column;
}

.host h1 {
    margin: 0 0 8px 0;
}

.host h4 {
    margin: 0 0 24px 0;
    font-weight: 400;
    color: #333;
}

.card {
    display: flex;
    border: 1px solid #c2c2c2;
    border-radius: 4px;
    padding: 24px 32px;
    flex-direction: column;
}
