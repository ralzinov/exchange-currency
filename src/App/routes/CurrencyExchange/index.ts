import * as React from 'react';

export default React.lazy(() => {
    import('./store');
    return import('./CurrencyExchange');
});
