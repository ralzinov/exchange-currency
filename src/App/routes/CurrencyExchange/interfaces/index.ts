export * from './IWallet';
export * from './ICurrencyExchangeReducerState';
export * from './ICurrencyExchangeProps';
