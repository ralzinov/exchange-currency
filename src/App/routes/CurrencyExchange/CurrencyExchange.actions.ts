import {IActionCreator, ISideEffectActionCreator} from '../../store/interfaces';
import {IWallet} from './interfaces';


export const LOAD_CURRENCY_RATES = '[Currency] Load currency rate';
export const LOAD_CURRENCY_RATES_SUCCESS = '[Currency] Currency rate loaded';
export const LOAD_CURRENCY_RATES_ERROR = '[Currency] Failed to load currency rate';

export const LOAD_WALLETS = '[Currency] Load user wallets';
export const LOAD_WALLETS_ERROR = '[Currency] Failed to load user wallets';
export const LOAD_WALLETS_SUCCESS = '[Currency] User wallets loaded';
export const AMOUNT_FIELD_CHANGE = '[Currency form] Amount field change';

const loadCurrencyRatesSuccess: IActionCreator<any> = () => ({
    type: LOAD_CURRENCY_RATES_SUCCESS
});

const loadCurrencyRatesError: IActionCreator<any> = () => ({
    type: LOAD_CURRENCY_RATES_ERROR
});

export const loadCurrencyRates: ISideEffectActionCreator<{base: string; target: string}> = (payload) => ({
    onSuccess: loadCurrencyRatesSuccess,
    onError: loadCurrencyRatesError,
    type: LOAD_CURRENCY_RATES,
    payload
});


const loadWalletsSuccess: IActionCreator<IWallet[]> = (payload) => ({
    type: LOAD_WALLETS_SUCCESS,
    payload
});

const loadWalletsError: IActionCreator<{}> = (payload) => ({
    type: LOAD_WALLETS_ERROR,
    payload
});

export const loadWallets: ISideEffectActionCreator = () => ({
    onSuccess: loadWalletsSuccess,
    onError: loadWalletsError,
    type: LOAD_WALLETS
});

export const amountFieldChange: IActionCreator<string> = (payload) => ({
    type: AMOUNT_FIELD_CHANGE,
    payload
});


export type ICurrencyRatesActions =
    ReturnType<typeof amountFieldChange> |
    ReturnType<typeof loadCurrencyRates> |
    ReturnType<typeof loadWallets>;
