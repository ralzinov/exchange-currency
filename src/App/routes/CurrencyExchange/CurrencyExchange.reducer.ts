import {createSelector} from 'reselect';
import {formValueSelector} from 'redux-form';
import {IAPIErrorResponse, ICurrencyExchangeReducerState, ICurrencyPairExchangeRate, IWallet} from './interfaces';
import {
    EXCHANGE_RATE_UPDATE_ERROR,
    EXCHANGE_RATE_UPDATE_SUCCESS,
    ICurrencyRatesActions,
    LOAD_WALLETS,
    LOAD_WALLETS_ERROR,
    LOAD_WALLETS_SUCCESS,
    RUN_EXCHANGE_RATE_POLLING,
    STOP_EXCHANGE_RATE_POLLING
} from './CurrencyExchange.actions';
import {IFormSelectOption} from '../../../components/FormSelect';
import {memoize} from '../../../utils';

export const EXCHANGE_FORM_NAME = 'exchange';

const initialState: ICurrencyExchangeReducerState = {
    isWalletsLoadPending: false,
    isExchangeRatePollingEnabled: false,
    wallets: []
};

export function currencyExchangeReducer(
    state: ICurrencyExchangeReducerState = initialState,
    action: ICurrencyRatesActions
): ICurrencyExchangeReducerState {
    switch (action.type) {
        case LOAD_WALLETS: {
            return {
                ...state,
                isWalletsLoadPending: true,
            };
        }

        case LOAD_WALLETS_SUCCESS: {
            return {
                ...state,
                isWalletsLoadPending: false,
                wallets: action.payload as IWallet[]
            };
        }

        case LOAD_WALLETS_ERROR: {
            return {
                ...state,
                walletsLoadError: action.payload as IAPIErrorResponse,
                isWalletsLoadPending: false,
            };
        }

        case EXCHANGE_RATE_UPDATE_SUCCESS: {
            return {
                ...state,
                exchangeRate: action.payload as ICurrencyPairExchangeRate
            };
        }

        case EXCHANGE_RATE_UPDATE_ERROR: {
            return {
                ...state,
                exchangeRateError: action.payload as IAPIErrorResponse,
                isExchangeRatePollingEnabled: false
            };
        }

        case RUN_EXCHANGE_RATE_POLLING: {
            return {
                ...state,
                isExchangeRatePollingEnabled: true
            };
        }

        case STOP_EXCHANGE_RATE_POLLING: {
            return {
                ...state,
                isExchangeRatePollingEnabled: false
            };
        }

        default: return state;
    }
}

const EXCHANGE_STATE_SELECTOR = (state: { exchange: ICurrencyExchangeReducerState }) => state.exchange;

export const EXCHANGE_POLLING_ENABLED_SELECTOR = createSelector(
    EXCHANGE_STATE_SELECTOR,
    (exchange) => {
        return exchange && exchange.isExchangeRatePollingEnabled;
    }
);

export const WALLETS_EXCHANGE_RATES_SELECTOR = createSelector(
    EXCHANGE_STATE_SELECTOR,
    (exchange) => exchange && exchange.exchangeRate && exchange.exchangeRate.rates
);

export const FORM_VALUE_SELECTOR_FACTORY = (field: string) => (state: object) => {
    return formValueSelector(EXCHANGE_FORM_NAME)(state, field);
};

const mapWalletToOption = memoize((wallet: IWallet): IFormSelectOption => {
    return {
        label: wallet.currency,
        value: wallet
    };
});

const WALLETS_SELECTOR = createSelector(
    EXCHANGE_STATE_SELECTOR,
    (exchange) => exchange && exchange.wallets
);

export const WALLETS_OPTIONS_SELECTOR_FACTORY = (fieldName: string) => createSelector(
    WALLETS_SELECTOR,
    FORM_VALUE_SELECTOR_FACTORY(fieldName),
    (wallets, source: IWallet) => {
        return wallets
            .filter((wallet) => wallet !== source)
            .map(mapWalletToOption);
    }
);

