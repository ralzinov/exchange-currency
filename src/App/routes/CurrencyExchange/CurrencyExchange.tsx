import * as React from 'react';
import styles from './CurrencyExchange.module.css'
import CurrencyExchangeForm from './CurrencyExchangeForm';
import {ICurrencyExchangeProps} from './interfaces';
import {useEffect} from 'react';

const submitForm = (props: any) => {
    console.log(props);
};

const CurrencyExchange: React.FC<ICurrencyExchangeProps> = (props) => {
    useEffect(() => void props.loadWallets(), []);

    return (
        <div className={styles.host}>
            <h1>Transfer money</h1>
            <h4>Transfer money between wallets</h4>
            <div className={styles.card}>
                <CurrencyExchangeForm onSubmit={submitForm}/>
            </div>
        </div>
    );
};

export default CurrencyExchange;

