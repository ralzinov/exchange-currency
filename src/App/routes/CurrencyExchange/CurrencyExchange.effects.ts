import {fromFetch} from 'rxjs/fetch';
import {combineLatest, empty, from, fromEvent, merge, of, throwError, timer} from 'rxjs';
import {ActionsObservable, ofType, StateObservable} from 'redux-observable';
import {
    catchError,
    exhaustMap,
    filter,
    map,
    mapTo,
    startWith,
    switchMap,
    take,
    takeUntil,
    withLatestFrom
} from 'rxjs/operators';
import {
    EXCHANGE_RATE_UPDATE_ERROR,
    LOAD_WALLETS,
    loadWallets,
    PAUSE_EXCHANGE_RATE_POLLING,
    pauseExchangeRatePolling,
    RESUME_EXCHANGE_RATE_POLLING,
    resumeExchangeRatePolling,
    RUN_EXCHANGE_RATE_POLLING,
    runExchangeRatePolling,
    STOP_EXCHANGE_RATE_POLLING,
    stopExchangeRatePolling
} from './CurrencyExchange.actions';
import {FORM_VALUE_SELECTOR_FACTORY} from './CurrencyExchange.reducer';
import {ICurrencyExchangeAPIResponse, IWallet} from './interfaces';
import {select} from '../../store';
import {Action} from 'redux';

const EXCHANGE_SERVER_URI = 'https://api.exchangerate-api.com/v4/latest/'; // TODO move to .env
const CURRENCY_EXCHANGE_POLLING_INTERVAL_MS = 1000;
const UNHANDLED_EXCEPTION_STATUS = -1;

const randomBalance = (decimals: number) => parseFloat((Math.random() * decimals * 10).toFixed(2));
const MOCK_WALLETS: IWallet[] = [{
    balance: randomBalance(4000),
    currency: 'USD',
}, {
    balance: randomBalance(10000),
    currency: 'RUB'
}, {
    balance: randomBalance(6000),
    currency: 'EUR'
}, {
    balance: randomBalance(3000),
    currency: 'GBP'
}];

export const loadWalletsEffect = (action$: ActionsObservable<ReturnType<typeof loadWallets>>) =>
    action$.pipe(
        ofType(LOAD_WALLETS),
        switchMap(({payload, onSuccess, onError}) => {
            // TODO Mocking API here
            // TODO make actual request
            // TODO handle error
            return timer(100).pipe(
                mapTo(onSuccess(MOCK_WALLETS))
            );
        })
    );

const formFieldValueFactory = (state$: StateObservable<object>, field: string) => {
    return select(state$, FORM_VALUE_SELECTOR_FACTORY(field)).pipe(
        filter((value) => Boolean(value)),
        map(({value}) => value.currency)
    );
};

const exchangeRatesRequestFactory = ({onSuccess, onError}: ReturnType<typeof runExchangeRatePolling>) =>
    (source: IWallet, target: IWallet) => fromFetch(`${EXCHANGE_SERVER_URI}${source}`).pipe(
        switchMap((response) => response.ok ? from(response.json()) : throwError(response)),
        map((data: ICurrencyExchangeAPIResponse) => {
            return onSuccess({
                timestamp: data.time_last_updated,
                rates: data.rates,
                base: data.base
            });
        }),
        catchError((response) => {
            return of(onError({
                status: response.status || UNHANDLED_EXCEPTION_STATUS,
                message: `Failed to convert ${source.currency} to ${target.currency}`
            }))
        })
    );

export const runExchangeRatePollingEpic = (
    action$: ActionsObservable<ReturnType<typeof runExchangeRatePolling>>,
    state$: StateObservable<object>
) => {
    const currencyPairValues$ = combineLatest(
        formFieldValueFactory(state$, 'source'),
        formFieldValueFactory(state$, 'target'),
    );

    return action$.pipe(ofType(RUN_EXCHANGE_RATE_POLLING)).pipe(
        switchMap((runAction) => {
            const request$ = exchangeRatesRequestFactory(runAction);
            return action$.pipe(
                ofType(PAUSE_EXCHANGE_RATE_POLLING, RESUME_EXCHANGE_RATE_POLLING),
                startWith(runAction),
                switchMap((action) => {
                    console.log(action);
                    if (action.type === PAUSE_EXCHANGE_RATE_POLLING) {
                        return empty();
                    }
                    return timer(0, CURRENCY_EXCHANGE_POLLING_INTERVAL_MS).pipe(
                        withLatestFrom(currencyPairValues$),
                        exhaustMap(([, [source, target]]) => request$(source, target)),
                        takeUntil(action$.pipe(ofType(STOP_EXCHANGE_RATE_POLLING)))
                    );
                })
            );
        })
    );
};

export const pauseExchangeRatePollingOnOfflineEpic = (action$: ActionsObservable<Action>) => {
    return action$.pipe(
        ofType(RUN_EXCHANGE_RATE_POLLING, RESUME_EXCHANGE_RATE_POLLING),
        switchMap((() => merge(
            of(navigator.onLine),
            fromEvent(window, 'offline').pipe(mapTo(false))
        ).pipe(
            filter((isOnline) => !isOnline),
            map(() => pauseExchangeRatePolling()),
            takeUntil(action$.pipe(ofType(STOP_EXCHANGE_RATE_POLLING))),
            take(1)
        )))
    );
};

export const resumeExchangeRatePollingOnOnlineEpic = (action$: ActionsObservable<Action>) => {
    return action$.pipe(
        ofType(PAUSE_EXCHANGE_RATE_POLLING),
        switchMap((() => merge(
            of(navigator.onLine),
            fromEvent(window, 'online').pipe(mapTo(true))
        ).pipe(
            filter(Boolean),
            map(() => resumeExchangeRatePolling()),
            takeUntil(action$.pipe(ofType(STOP_EXCHANGE_RATE_POLLING, RUN_EXCHANGE_RATE_POLLING))),
            take(1)
        )))
    );
};

export const stopExchangeRatePollingOnErrorEpic = (action$: ActionsObservable<Action>) => {
    return action$.pipe(
        ofType(EXCHANGE_RATE_UPDATE_ERROR),
        mapTo(stopExchangeRatePolling())
    );
};

