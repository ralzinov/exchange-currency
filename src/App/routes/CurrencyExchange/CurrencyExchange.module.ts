import {connect} from 'react-redux';
import {ReducerRegistry} from '../../store/ReducerRegistry';
import {currencyExchangeReducer} from './CurrencyExchange.reducer';
import {loadWallets} from './CurrencyExchange.actions';
import {ICurrencyExchangeProps} from './interfaces';
import CurrencyExchange from './CurrencyExchange';

ReducerRegistry.register({
    exchange: currencyExchangeReducer
});

export default connect(null, {
    loadWallets
} as ICurrencyExchangeProps)(CurrencyExchange);
