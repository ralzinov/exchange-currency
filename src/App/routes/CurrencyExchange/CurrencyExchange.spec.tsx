import React from 'react';
import ReactDOM from 'react-dom';
import CurrencyExchange from './CurrencyExchange';
import {create} from 'react-test-renderer';
import {ICurrencyExchangeProps} from './interfaces';
import {act} from 'react-dom/test-utils';
// import {mount} from 'enzyme';

jest.mock('./CurrencyExchangeForm', () => ({options, value, onChange,}: any) => {
    const React = require('react');
    return <form/>
});

it('renders without crashing', () => {
    const props = {
        loadWallets: () => {
        }
    } as ICurrencyExchangeProps;
    const div = document.createElement('div');
    ReactDOM.render(<CurrencyExchange {...props}/>, div);
    ReactDOM.unmountComponentAtNode(div);
});

it('should render with props', () => {
    const props = {
        loadWallets: () => {
        }
    } as ICurrencyExchangeProps;
    const button = create(
        <CurrencyExchange {...props}/>
    );
    expect(button.toJSON()).toMatchSnapshot();
});

it('should load wallets after mount', () => {
    const div = document.createElement('div');
    const props = {
        loadWallets: jest.fn()
    } as ICurrencyExchangeProps;

    expect(props.loadWallets).not.toHaveBeenCalled();
    act(() => {
        ReactDOM.render(<CurrencyExchange {...props}/>, div);
    });
    expect(props.loadWallets).toHaveBeenCalled();
    ReactDOM.unmountComponentAtNode(div);
});

it('should ... on form submit', () => {

});

