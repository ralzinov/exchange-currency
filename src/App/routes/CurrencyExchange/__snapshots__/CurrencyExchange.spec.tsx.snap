// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with props 1`] = `
<div
  className="host"
>
  <h1>
    Transfer money
  </h1>
  <h4>
    Transfer money between wallets
  </h4>
  <div
    className="card"
  >
    <form />
  </div>
</div>
`;
